src/Components/Components/src/Rendering/RenderTreeUpdater.cs

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Components.RenderTree;
+
+namespace Microsoft.AspNetCore.Components.Rendering
+{
+    internal class RenderTreeUpdater
+    {
+        public static void UpdateToMatchClientState(RenderTreeBuilder renderTreeBuilder, int eventHandlerId, object newFieldValue)
+        {
+            // We only allow the client to supply string or bool currently, since those are the only kinds of
+            // values we output on attributes that go to the client
+            if (!(newFieldValue is string || newFieldValue is bool))
```

Comment:
In which case can this not be either an int or bool, wouldn't that signal that something went terribly wrong? (Like could a number get in here somehow?)

Reply:
Good point.

I'm not aware of any valid way that the value here might be something other than `string`/`bool`. If it was, we could just throw. Maybe that would be clearer.

Reply:
Looking at your tests, it's fine if this simply returns as the value comes from the client and is untrusted

Reply:
This might help a bit - https://github.com/aspnet/AspNetCore/issues/11517